NKL/Encode.lean

```diff
@@ -151,117 +150,202 @@ private def decList (f : DecodeM a) : DecodeM (List a) := do
+-- Affine Expressions
+
+def encIndexExpr : IndexExpr -> ByteArray
+  | .var name  => tag 0x10 [encString name]
+  | .int i     => tag 0x11 [encInt i]
+  | .neg e     => tag 0x12 [encIndexExpr e]
+  | .add l r   => tag 0x13 [encIndexExpr l, encIndexExpr r]
+  | .mul i e   => tag 0x14 [encInt i, encIndexExpr e]
+  | .floor e i => tag 0x15 [encIndexExpr e, encInt i]
+  | .ceil e i  => tag 0x16 [encIndexExpr e, encInt i]
+  | .mod e i   => tag 0x17 [encIndexExpr e, encInt i]
+
+partial def decIndexExpr : DecodeM IndexExpr := do
+  match (<- next) with
+  | 0x10 => return .var (<- decString)
+  | 0x11 => return .int (<- decInt)
+  | 0x12 => return .neg (<- decIndexExpr)
+  | 0x13 => return .add (<- decIndexExpr) (<- decIndexExpr)
+  | 0x14 => return .mul (<- decInt) (<- decIndexExpr)
+  | 0x15 => return .floor (<- decIndexExpr) (<- decInt)
+  | 0x16 => return .ceil (<- decIndexExpr) (<- decInt)
+  | 0x17 => return .mod (<- decIndexExpr) (<- decInt)
+  | t    => throw s!"Unknown tag in IndexExpr {t}"
+
+private def chkIE (e: IndexExpr) : Bool :=
+  (decode' decIndexExpr $ encIndexExpr e) == some e
+
+private def ie_var : IndexExpr := .var "s"
+
+#guard chkIE (.var "v")
+#guard chkIE (.int 1)
+#guard chkIE (.neg ie_var)
+#guard chkIE (.add ie_var ie_var)
+#guard chkIE (.mul 2 ie_var)
+#guard chkIE (.floor ie_var 2)
+#guard chkIE (.ceil ie_var 2)
+#guard chkIE (.mod ie_var 2)
+
+def encIndex : Index -> ByteArray
+  | .ellipsis    => tag 0x20 []
+  | .coord e     => tag 0x21 [enc e]
+  | .range l u s => tag 0x22 [enc l, enc u, enc s]
+where
+  enc := encOption encIndexExpr
+
+def decIndex : DecodeM Index := do
+  match (<- next) with
+  | 0x20 => return .ellipsis
+  | 0x21 => return .coord (<- dec)
+  | 0x22 => return .range (<- dec) (<- dec) (<- dec)
+  | t    => throw s!"Unknown tag in Index {t}"
+where
+  dec:= decOption decIndexExpr
+
+private def chkIndex (i : Index) : Bool :=
+  (decode' decIndex $ encIndex i) == some i
+
+#guard chkIndex .ellipsis
+#guard chkIndex (.coord none)
+#guard chkIndex (.coord $ some ie_var)
+#guard chkIndex (.range (some ie_var) none none)
 
 ------------------------------------------------------------------------------
 -- Expressions
 
 partial def encExpr : Expr -> ByteArray
-  | .value c          => tag 0x10 [encConst c]
-  | .bvar s           => tag 0x11 [encString s]
-  | .var s _          => tag 0x12 [encString s]
-  | .subscript e ix   => tag 0x13 [encExpr e, encList encExpr ix]
-  | .slice l u step   => tag 0x14 [encExpr l, encExpr u, encExpr step]
-  | .binop op l r     => tag 0x15 [encString op, encExpr l, encExpr r]
-  | .cond c t e       => tag 0x16 [encExpr c, encExpr t, encExpr e]
-  | .tuple es         => tag 0x17 [encList encExpr es]
-  | .list es          => tag 0x18 [encList encExpr es]
-  | .call f ax        => tag 0x19 [encExpr f, encList encExpr ax]
-  | .gridcall f ix ax => tag 0x1a [encExpr f, encList encExpr ix, encList encExpr ax]
```

Comment:
what happened to gridcall?

Reply:
Likely it will have to be added back, depending on what we do in the tracing.